### PR TITLE
[WIP]Support dynamic graph distributed

### PR DESCRIPTION
**Pull request by ForFishes**
### PR types

 New features
### PR changes

Others
### Describe

1、Support dynamic graph distributed by group design.

### CODE REVIEW

File: paddle/fluid/imperative/all_reduce.cc [hunk 1]
```diff
@@ -191,6 +191,47 @@ void AllReduce(const framework::Variable &src, framework::Variable *dst,
   AllReduce(src, dst, strategy, stream);
 }
 
+void AllReduce(
```
Comment (hutuxian) [inline paddle/fluid/imperative/all_reduce.cc]: We don't need these, right?

File: paddle/fluid/imperative/all_reduce.cc [hunk 2]
```diff
@@ -191,6 +191,47 @@ void AllReduce(const framework::Variable &src, framework::Variable *dst,
...
+    const std::unique_ptr<paddle::platform::CUDADeviceContext> &ctx) {
+  const auto strategy = ParallelStrategy();
+  cudaStream_t stream = ctx->stream();
+  AllReduce(src, dst, strategy, stream);
+}
+
+void SyncCommStream(
+    const std::unique_ptr<paddle::platform::CUDADeviceContext> &ctx) {
+  auto stream = ctx->stream();
+  PADDLE_ENFORCE_CUDA_SUCCESS(cudaStreamSynchronize(stream));
```
Comment (hutuxian) [inline paddle/fluid/imperative/all_reduce.cc]: call its Wait member function?

File: python/paddle/fluid/tests/unittests/test_imperative_group.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        self._linear1 = Linear(784, 10)
+        self._linear2 = Linear(10, 10)
+
+    def forward(self, inputs):
+        y = self._linear1(inputs)
+        y = self._linear2(y)
+        return y
+
+
+class TestDataParallelBucket(unittest.TestCase):
```
Comment (hutuxian) [inline python/paddle/fluid/tests/unittests/test_imperative_group.py]: TestDataParallelGroup?

File: paddle/fluid/imperative/reducer.cc [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+        const std::string &var_name = vars_[index]->GradVarName();
+        varname2index_[var_name] = VariableIndex{
+            .group_index = group_index, .variable_index = var_index,
+        };
+      }
+    }
+  }
+}
+
+void Reducer::initialize_groups(
```
Comment (hutuxian) [inline paddle/fluid/imperative/reducer.cc]: function name should comply with the conventions: (InitializeGroups here), the same as following several functions

File: paddle/fluid/imperative/all_reduce.cc (hunk 2 above)
Comment (ForFishes) [reply paddle/fluid/imperative/all_reduce.cc]: done

File: python/paddle/fluid/tests/unittests/test_imperative_group.py (hunk 3 above)
Comment (ForFishes) [reply python/paddle/fluid/tests/unittests/test_imperative_group.py]: done

File: paddle/fluid/imperative/reducer.cc (hunk 4 above)
Comment (ForFishes) [reply paddle/fluid/imperative/reducer.cc]: ok,fix it

File: paddle/fluid/imperative/all_reduce.cc (hunk 1 above)
Comment (ForFishes) [reply paddle/fluid/imperative/all_reduce.cc]: yes, will remove it

**Pull request closed without merging**